File: algorithms/95_UniqueBinarySearchTreesII.py
r# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def generateTrees(self, n):
        if not n: return []
        l = [i+1 for i in range(n)]
        res = self.dfs(l)
        return res
    
    def dfs(self, remaining_arr):
        if not remaining_arr:
            return [None]
        
        trees = []
        for i in range(len(remaining_arr)):
            left_trees = self.dfs(remaining_arr[:i])
            right_trees = self.dfs(remaining_arr[i+1:])
            for l in left_trees:
                for r in right_trees:
                    root = TreeNode(remaining_arr[i])
                    root.left = l
                    root.right = r
                    trees.append(root)
        return trees

File: algorithms/95_UniqueBinarySearchTreesII.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public List<TreeNode> generateTrees(int n) {
        if (n==0) { return new LinkedList<TreeNode>(); }
        int[] arr = new int[n];
        for (int i=0; i<n; ++i) {
            arr[i] = i+1;
        }
        return dfs(arr);
    }
    
    public List<TreeNode> dfs(int[] remaining) {
        List<TreeNode> trees = new LinkedList<TreeNode>();
        if (remaining.length == 0) {
            trees.add(null);
            return trees;
        }
        
        for (int i=0; i< remaining.length; ++i) {
            List<TreeNode> leftChilds = dfs(Arrays.copyOfRange(remaining, 0, i));
            List<TreeNode> rightChilds = dfs(Arrays.copyOfRange(remaining, i+1, remaining.length));
            for (TreeNode l: leftChilds) {
                for (TreeNode r: rightChilds) {
                    TreeNode root = new TreeNode(remaining[i]);
                    root.left = l;
                    root.right = r;
                    trees.add(root);
                }
            }
        }
        return trees;
    }
}